cvat-core/src/requests-manager.ts

```diff
@@ -0,0 +1,357 @@
+        }
+        if (request.message !== undefined) {
+            this.#message = request.message;
+        }
+        if (request.result_url !== undefined) {
+            this.#resultUrl = request.result_url;
+        }
+        if (request.enqueued_date !== undefined) {
+            this.#enqueuedDate = request.enqueued_date;
+        }
+        if (request.started_date !== undefined) {
+            this.#startedDate = request.started_date;
+        }
+        if (request.finished_date !== undefined) {
+            this.#finishedDate = request.finished_date;
+        }
+        if (request.expiry_date !== undefined) {
+            this.#expiryDate = request.expiry_date;
+        }
+        if (request.owner !== undefined) {
+            this.#owner = new User(request.owner);
+        }
+    }
+}
+
+const REQUESTS_COUNT = 5;
+const PROGRESS_EPS = 25;
+const EPS_DELAYS = {
+    [RQStatus.STARTED]: [3000, 7000, 13000],
+    [RQStatus.QUEUED]: [7000, 13000, 19000],
+    [RQStatus.DEFERRED]: [120000, 180000, 240000],
+};
+
+class RequestsManager {
+    private listening: Record<string, {
+        onUpdate: ((request: Request) => void)[];
+        requestDelayIdx: number | null,
+        request: Request | null,
+        timeout: number | null;
+        promise?: Promise<Request>;
+    }>;
+
+    private requestStack: number[];
+    constructor() {
+        this.listening = {};
+        this.requestStack = [];
+    }
+
+    async list(): Promise<{ requests: Request[], count: number }> {
+        const result = await serverProxy.requests.list();
+        const requests = [];
+        for (const request of result) {
+            requests.push(
+                new Request({
+                    ...request,
+                }),
+            );
+        }
+        return { requests, count: requests.length };
+    }
+
+    async listen(
+        id: string | null,
+        options?: {
+            callback?: (request: Request) => void,
+            filter?: RequestsFilter,
+            initialRequest?: Request,
+        },
+    ): Promise<Request> {
+        let storedID = null;
+        const params = options?.filter ? fieldsToSnakeCase(options.filter) : {};
+        if (id) {
+            storedID = id;
+        } else if (options?.filter) {
+            storedID = new URLSearchParams(params).toString();
+        }
+
+        if (!storedID) {
+            throw new Error('Neither request id nor filter is provided');
+        }
+        const callback = options?.callback;
+        const initialRequest = options?.initialRequest;
+
+        if (storedID in this.listening) {
+            if (callback) {
+                this.listening[storedID].onUpdate.push(callback);
+            }
+            return this.listening[storedID].promise;
+        }
+        const promise = new Promise<Request>((resolve, reject) => {
+            const timeoutCallback = (): void => {
+                if (!(storedID in this.listening)) {
+                    return;
```

Comment:
How is that possible if you clear timeout before deleting the request from this.listening?

Reply:
Removed